### PR TITLE
HWY-188: Introducing Tests for HighwayProtocol

### PR DESCRIPTION
**Pull request by xcthulhu**
JIRA: https://casperlabs.atlassian.net/browse/HWY-188

### CODE REVIEW

**Comment by goral09** [comment]
@xcthulhu Can you add a bit more to the description of the PR? I don't understand what kind of tests you are introducing. Is this overlapping with @afck 's latest tests of the `MockProto`?

**Comment by afck** [comment]
> Is this overlapping with @afck 's latest tests of the `MockProto`?

No, `MockProto` _replaces_ `HighwayProtocol` in tests.

**Comment by goral09** [comment]
> > Is this overlapping with @afck 's latest tests of the `MockProto`?
> 
> No, `MockProto` _replaces_ `HighwayProtocol` in tests.

And this PR starts testing `HighwayProtocol` directly. Won't tests of `HighwayProtocol` overlap with tests based on `MockProto`?

File: node/src/components/consensus/protocols/highway/tests.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    match &opt_protocol_outcome {
+        None => panic!("We just checked that effects has length 1!"),
+        Some(ProtocolOutcome::InvalidIncomingMessage(invalid_msg, offending_sender, _err)) => {
+            assert_eq!(
+                invalid_msg, &msg,
+                "Invalid message is not message that was sent.\
+                  Invalid message in protocol outcome: {:?} ;\
+                  Message that was sent: {:?}",
+                invalid_msg, &msg
```
Comment (afck) [inline node/src/components/consensus/protocols/highway/tests.rs:80]: Aren't those two printed anyway? I think these three lines can be omitted.

Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/tests.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/536/commits/8600a710ebccb65588a46d78fad7f1c6ada1fa9c

**Comment by xcthulhu** [comment]
> Won't tests of `HighwayProtocol` overlap with tests based on `MockProto`?

Perhaps.  I introduced this infrastructure because I ultimately need to test the UML diagrams in [CEP-19][1]. I got the impression I need to test `HighwayProtocol`.

Now, even if this is the wrong way to go about it, doesn't this test increase our coverage?  Before this `HighwayProtocol` wasn't tested as far as I could tell...

[1]: https://github.com/CasperLabs/ceps/blob/master/text/0019-fork-spam.md#guide-level-explanation

**Comment by goral09** [comment]
> Now, even if this is the wrong way to go about it, doesn't this test increase our coverage? Before this `HighwayProtocol` wasn't tested as far as I could tell...

You're right, it wasn't. Andreas introduced `MockProto` to not have to interact with `HighwayProtocol` directly. @afck care to weigh in?

**Comment by afck** [comment]
Exactly: `MockProto` is to test the `EraSupervisor` without having to also deal with the complicated details of `HighwayProtocol`.
This test here is precisely about testing `HighwayProtocol` — and without the era supervisor!

**Review by goral09 — APPROVED** [review]
I guess this PR does what it's supposed to do – add some ground work to test `HighwayProtocol`.

**Pull request merged**